### PR TITLE
MNT: Add and apply pre-commit configuration

### LINKED ISSUE #1168: Integrate pre-commit
Pre-commit hooks will enable auto-formatting (see #1164) and other beneficial actions/validations.

### PR DESCRIPTION
**Pull request by effigies**
Closes #1168.

### CODE REVIEW

File: doc/source/dicom/spm_dicom.rst [hunk 1]
```diff
@@ -269,7 +269,7 @@ Then define the following matrices:
 .. math::
 
    R = \left(\begin{smallmatrix}1 & a & 1 & 0\\1 & b & 0 & 1\\1 & c & 0 & 0\\1 & d & 0 & 0\end{smallmatrix}\right)
-   
+
```
Comment (effigies) [inline doc/source/dicom/spm_dicom.rst:272]: Is this whitespace significant?

Comment (ZviBaratz) [reply doc/source/dicom/spm_dicom.rst:272]: No, I assume it's just the `end-of-file-fixer` hook.

Comment (effigies) [reply doc/source/dicom/spm_dicom.rst:272]: Right. I just meant we should verify that this doesn't change the appearance (or the change is acceptable) of the rendered doc.

File: nibabel/gifti/tests/data/ascii.gii [hunk 2]
```diff
@@ -48,15 +48,15 @@
...
          <TransformedSpace><![CDATA[NIFTI_XFORM_TALAIRACH]]></TransformedSpace>
          <MatrixData>
-            1.000000 0.000000 0.000000 0.000000 
-            0.000000 1.000000 0.000000 0.000000 
-            0.000000 0.000000 1.000000 0.000000 
-            0.000000 0.000000 0.000000 1.000000 
+            1.000000 0.000000 0.000000 0.000000
+            0.000000 1.000000 0.000000 0.000000
+            0.000000 0.000000 1.000000 0.000000
+            0.000000 0.000000 0.000000 1.000000
```
Comment (effigies) [inline nibabel/gifti/tests/data/ascii.gii:54]: While some of these are probably semantically null changes, I think it's bad policy to automatically change a data file. We should cordon off `data/` directories from being inspected/modified by linters.

File: doc/source/dicom/spm_dicom.rst (hunk 1 above)
Comment (ZviBaratz) [reply doc/source/dicom/spm_dicom.rst:272]: :sweat_smile: :sweat_smile: 
Just verified, no change in appearance.

File: nibabel/gifti/tests/data/ascii.gii (hunk 2 above)
Comment (ZviBaratz) [reply nibabel/gifti/tests/data/ascii.gii]: Reverted changes to data directories and added `'.*/data/.*'` to the pre-commit configuration top-level exclusion regex :+1:

**Comment by effigies** [comment]
LGTM. I broke it up so that the configuration changes are not in the same commits as the style changes. Please merge if you're happy with this.

**Comment by ZviBaratz** [comment]
Thank you! I'll work on improving the granularity of my commits.

**Pull request merged**